src/ultk/language/grammar/inference.py

```diff
@@ -0,0 +1,84 @@
+from typing import TypeVar, Iterable, Callable
+from ultk.language.grammar.grammar import Grammar, GrammaticalExpression
+from ultk.language.grammar.likelihood import Dataset, all_or_nothing
+import copy
+import random
+
+
+def mh_sample(
+    expr: GrammaticalExpression,
+    grammar: Grammar,
+    data: Dataset,
+    likelihood_func: Callable[[Dataset, GrammaticalExpression], float] = all_or_nothing,
+) -> GrammaticalExpression:
+    old_tree_prior = grammar.prior(expr)
+    old_node_count = expr.node_count()
+    while True:
+        old_tree = copy.deepcopy(expr)
+        current_node, parent_node = mh_select(old_tree)
+        old_subtree_prior = grammar.prior(current_node)
+        new_tree, new_node = mh_generate(old_tree, current_node, parent_node, grammar)
+        new_tree_prior = grammar.prior(new_tree)
+        new_node_count = new_tree.node_count()
+        new_subtree_prior = grammar.prior(new_node)
+        try:
+            mh_accept = min(
+                1,
+                (
+                    (new_tree_prior * likelihood_func(data, new_tree))
+                    / (old_tree_prior * likelihood_func(data, old_tree))
+                )
+                * (
+                    (old_subtree_prior / new_node_count)
+                    / (new_subtree_prior / old_node_count)
+                ),
+            )
+        except ZeroDivisionError:
+            mh_accept = 0
+        if random.random() < mh_accept:
+            return new_tree
+
+
+def mh_select(
+    old_tree: GrammaticalExpression,
+) -> (GrammaticalExpression, GrammaticalExpression):
```

Comment:
Return type should be `tuple[GrammaticalExpression, GrammaticalExpression]` (same for below too)